src/aggregation/metric/top_hits.rs

```diff
@@ -0,0 +1,813 @@
+                        value: Some(5),
+                        order: Order::Asc,
+                    }],
+                    Default::default(),
+                ),
+            },
+        ];
+
+        let mut collector = collector_with_capacity(3);
+        for doc in docs.clone() {
+            collector.collect(doc.feature, doc.doc);
+        }
+
+        let res = collector.finalize();
+
+        assert_eq!(
+            res,
+            super::TopHitsMetricResult {
+                hits: vec![
+                    super::TopHitsVecEntry {
+                        id: docs[0].doc,
+                        sort: vec![docs[0].feature.0[0].value],
+                        search_results: Default::default(),
+                    },
+                    super::TopHitsVecEntry {
+                        id: docs[1].doc,
+                        sort: vec![docs[1].feature.0[0].value],
+                        search_results: Default::default(),
+                    },
+                    super::TopHitsVecEntry {
+                        id: docs[2].doc,
+                        sort: vec![docs[2].feature.0[0].value],
+                        search_results: Default::default(),
+                    },
+                ]
+            }
+        );
+
+        Ok(())
+    }
+
+    fn test_aggregation_top_hits(merge_segments: bool) -> crate::Result<()> {
+        let docs = vec![
+            vec![
+                r#"{ "date": "2015-01-02T00:00:00Z", "text": "bbb", "text2": "bbb" }"#,
+                r#"{ "date": "2017-06-15T00:00:00Z", "text": "ccc", "text2": "ddd" }"#,
+            ],
+            vec![
+                r#"{ "text": "aaa", "text2": "bbb", "date": "2018-01-02T00:00:00Z" }"#,
+                r#"{ "text": "aaa", "text2": "bbb", "date": "2016-01-02T00:00:00Z" }"#,
+            ],
+        ];
+
+        let index = get_test_index_from_docs(merge_segments, &docs)?;
+
+        let d: Aggregations = serde_json::from_value(json!({
+            "top_hits_req": {
+                "top_hits": {
+                    "size": 2,
+                    "sort": [
+                        { "date": "desc" }
+                    ],
+                    "from": 1,
+                    "docvalue_fields": [
+                        "date",
+                        "tex*",
+                    ],
+                }
+        }
+        }))?;
+
+        let collector = AggregationCollector::from_aggs(d, Default::default());
+        let reader = index.reader()?;
+        let searcher = reader.searcher();
+
+        let agg_res =
+            serde_json::to_value(searcher.search(&AllQuery, &collector).unwrap()).unwrap();
+
+        let date_2017 = datetime!(2017-06-15 00:00:00 UTC);
+        let date_2016 = datetime!(2016-01-02 00:00:00 UTC);
+
+        assert_eq!(
+            agg_res["top_hits_req"],
+            json!({
+                "hits": [
+                    {
+                        "id": {
+                            "segment_ord": 0,
```

Comment:
why do you return the segment and docid here?

Reply:
Is there something else we can use to uniquely identify/retrieve documents?

Reply:
I mean do we do something with it or is it just for the user?

Reply:
Ah, yes it's for the user

Reply:
I think in the context of a searcher this is valid information, but outside to the user not so much (e.g. the segment ord may be deleted, or there are multiple indices)

Reply:
Makes sense. In the standard use-case for term-bucketing and retrieving top-x documents from each bucket, this information should be valid and useful, for example when the user wants to retrieve the entire document rather than a set of fast fields from each. On the other hand we should make it explicit that the identifier isn't useful outside of that context perhaps, how would you like it to be done?

Reply:
I think I would remove it and if there is some demand add more search field features like `docvalue fields`

Reply:
Our current use-case needs access to a large in-store text field, so it would be problematic if it is removed without an alternative like `fields` :/

Perhaps we can do a workaround where we exclude the DocAddress from the serialized form?

Alternatively I could also work on `fields` support with some help, but preferably in a follow-up PR sometime next month (exam time at uni!)

Reply:
Can you make it a fast field or would that be too expensive?
Otherwise I would remove it from the serialized final result, I think

Reply:
I'll evaluate making it a fast field before we proceed. Generally speaking, would  you draw the line considering 10,000-100,000 documents where this field can be 20-50KB? I'm not very familiar with how the term dictionary stores, loads and queries so I'm curious if there's an approach to estimating the memory expense.

Reply:
The term fastfield uses an sstable with compressed blocks. The blocks are relatively small, that means may we get many blocks. On the top of my head I'm not sure if that may cause problems, maybe @trinity-1686a knows more

Reply:
iirc, when building a block, only the size of the key portion of the table is taken into account for sizing blocks, not the value part. If I understand how fastfields uses the sstable, (key=docid, value=the entire field value, correct me if I'm mistaken), it means you'll get unusually large blocks, but they will each contain as many docs as they would if the values were smaller.

If I understand correctly everything, assuming doc_ids are sequential, we can store about 2k docs in a block, for an upper bound a little under 100MiB of raw data. So I'd say it would use in the order of 100MiB per loaded sstable at a given instant

Reply:
Thank you, that was helpful! I have removed DocAddress from top hits' results.